In [1]:
import pandas as pd

In [50]:
pd.read_excel

<function pandas.io.excel._base.read_excel(io, sheet_name: 'str | int | list[IntStrT] | None' = 0, header: 'int | Sequence[int] | None' = 0, names=None, index_col: 'int | Sequence[int] | None' = None, usecols=None, squeeze: 'bool | None' = None, dtype: 'DtypeArg | None' = None, engine: "Literal['xlrd', 'openpyxl', 'odf', 'pyxlsb'] | None" = None, converters=None, true_values: 'Iterable[Hashable] | None' = None, false_values: 'Iterable[Hashable] | None' = None, skiprows: 'Sequence[int] | int | Callable[[int], object] | None' = None, nrows: 'int | None' = None, na_values=None, keep_default_na: 'bool' = True, na_filter: 'bool' = True, verbose: 'bool' = False, parse_dates=False, date_parser=None, thousands: 'str | None' = None, decimal: 'str' = '.', comment: 'str | None' = None, skipfooter: 'int' = 0, convert_float: 'bool | None' = None, mangle_dupe_cols: 'bool' = True, storage_options: 'StorageOptions' = None) -> 'DataFrame | dict[IntStrT, DataFrame]'>

In [51]:
df = pd.read_excel(r'C:\Users\felip\Desktop\Workflow\Ciência de Dados\Projetos de Ciência de Dados com Python\default_of_credit_card_clients__courseware_version_1_21_19.xls')

In [52]:
print(df)

                  ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0      798fc410-45c1      20000    2          2         1   24     2      2   
1      8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2      85698822-43f5      90000    2          2         2   34     0      0   
3      0737c11b-be42      50000    2          2         1   37     0      0   
4      3b7f77cc-dbc0      50000    1          2         1   57    -1      0   
...              ...        ...  ...        ...       ...  ...   ...    ...   
29995  ecff42d0-bdc6     220000    1          3         1   39     0      0   
29996  99d1fa0e-222b     150000    1          3         2   43    -1     -1   
29997  95cdd3e7-4f24      30000    1          2         2   37     4      3   
29998  00d03f02-04cd      80000    1          3         1   41     1     -1   
29999  15d69f9f-5ad3      50000    1          2         1   46     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BIL

In [53]:
df.shape

(30000, 25)

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          30000 non-null  object
 1   LIMIT_BAL                   30000 non-null  int64 
 2   SEX                         30000 non-null  int64 
 3   EDUCATION                   30000 non-null  int64 
 4   MARRIAGE                    30000 non-null  int64 
 5   AGE                         30000 non-null  int64 
 6   PAY_1                       30000 non-null  object
 7   PAY_2                       30000 non-null  int64 
 8   PAY_3                       30000 non-null  int64 
 9   PAY_4                       30000 non-null  int64 
 10  PAY_5                       30000 non-null  int64 
 11  PAY_6                       30000 non-null  int64 
 12  BILL_AMT1                   30000 non-null  int64 
 13  BILL_AMT2                   30000 non-null  in

In [55]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

In [56]:
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1',
       'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2',
       'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1',
       'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

In [57]:
df['ID'].nunique()

29687

In [58]:
df.shape

(30000, 25)

In [59]:
id_counts = df['ID'].value_counts()

In [60]:
id_counts.head()

ad23fe5c-7b09    2
1fb3e3e6-a68d    2
89f8f447-fca8    2
7c9b7473-cc2f    2
90330d02-82d9    2
Name: ID, dtype: int64

In [61]:
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

Para **filtrar** e **limpar** os IDs duplicados ----> **MÁSCARA BOOLEANA**

In [62]:
import numpy as np

In [63]:
np.random.seed(seed=24)

In [64]:
random_integers = np.random.randint(low=1,high=5,size=100)

In [65]:
random_integers

array([3, 4, 1, 4, 2, 2, 2, 1, 4, 4, 1, 4, 4, 3, 4, 4, 4, 4, 4, 4, 2, 3,
       4, 4, 2, 4, 2, 4, 1, 1, 3, 1, 4, 2, 2, 1, 4, 3, 2, 2, 3, 3, 2, 3,
       4, 2, 3, 4, 1, 3, 4, 2, 2, 4, 1, 2, 3, 3, 2, 1, 2, 4, 2, 1, 4, 3,
       1, 2, 1, 3, 4, 2, 3, 4, 1, 3, 2, 1, 1, 3, 3, 4, 3, 2, 4, 1, 4, 4,
       4, 4, 1, 4, 3, 2, 4, 4, 3, 3, 2, 4])

In [66]:
is_equal_to_1 = random_integers ==1

In [67]:
is_equal_to_1

array([False, False,  True, False, False, False, False,  True, False,
       False,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True,  True, False,  True, False, False, False,  True,
       False, False, False, False, False, False, False, False, False,
       False, False, False,  True, False, False, False, False, False,
        True, False, False, False, False,  True, False, False, False,
        True, False, False,  True, False,  True, False, False, False,
       False, False,  True, False, False,  True,  True, False, False,
       False, False, False, False,  True, False, False, False, False,
        True, False, False, False, False, False, False, False, False,
       False])

In [68]:
sum(is_equal_to_1)

18

Agora que já transformamos os elementos que queríamos em booleans e conseguimos localizá-los e contá-los:

-------- vamos selecioná-los (**indexação de arrays**) utilizando a máscara booleana -------

SAINDO DO **EXEMPLO ALEATÓRIO** E VOLTANDO AO **DATAFRAME**

no exemplo nós queríamos os valores 1 do array

no dataframe vamos querer os valores da série(coluna) ------- id_counts ==2 -------

In [69]:
dupe_mask = id_counts ==2

In [70]:
dupe_mask

ad23fe5c-7b09     True
1fb3e3e6-a68d     True
89f8f447-fca8     True
7c9b7473-cc2f     True
90330d02-82d9     True
                 ...  
2edeb3a6-d38d    False
27e11893-92e9    False
2b1c282b-441e    False
bd737997-0eb0    False
15d69f9f-5ad3    False
Name: ID, Length: 29687, dtype: bool

Os IDs agora são os índices e os booleans os valores da série id_counts

In [71]:
id_counts.index[:]

Index(['ad23fe5c-7b09', '1fb3e3e6-a68d', '89f8f447-fca8', '7c9b7473-cc2f',
       '90330d02-82d9', '2a793ecf-05c6', '75938fec-e5ec', '7be61027-a493',
       'a3a5c0fc-fdd6', 'b44b81b2-7789',
       ...
       'ed32d9fa-fc4d', '0d3c491a-f31a', '43b86f14-5293', '9aea1c3d-6a89',
       'eecd60d0-e496', '2edeb3a6-d38d', '27e11893-92e9', '2b1c282b-441e',
       'bd737997-0eb0', '15d69f9f-5ad3'],
      dtype='object', length=29687)

**SELECIONE E ARMAZENE** os IDs duplicados(Trues) em uma variável

In [72]:
dupe_ids = id_counts.index[dupe_mask]

Converta a variável para uma lista

In [73]:
dupe_ids = list(dupe_ids)

In [74]:
len(dupe_ids)

313

In [75]:
dupe_ids

['ad23fe5c-7b09',
 '1fb3e3e6-a68d',
 '89f8f447-fca8',
 '7c9b7473-cc2f',
 '90330d02-82d9',
 '2a793ecf-05c6',
 '75938fec-e5ec',
 '7be61027-a493',
 'a3a5c0fc-fdd6',
 'b44b81b2-7789',
 '998fa9b2-b341',
 '69566a6b-6156',
 '4e2380e6-a8cf',
 'b87bf8f3-d704',
 '4f95b36b-ab10',
 '214fc1ca-b25a',
 'ab09b4b4-8331',
 '4653bc48-0141',
 '3d832783-1571',
 'd9dafe9c-b497',
 '30729bb5-55d7',
 '26d4a6eb-8abe',
 '6c0797cd-f663',
 'e77f7765-8731',
 '3cd5c728-df14',
 'c361e43d-f1f4',
 '7e7a4087-4def',
 'ddcc7278-7017',
 'edd737fd-5b07',
 'cfa9d026-de0d',
 '6870544f-459e',
 '9ac5d94f-ed71',
 '2458b96d-e8f1',
 'bf23c70b-bf12',
 '3327c2a8-3402',
 '7fe414ae-32f8',
 '8d3a2576-a958',
 '82da3aa5-97a6',
 '45cceda0-6fb7',
 '9db1809e-845a',
 '3e60a98c-e520',
 'a17f0f1f-d9f6',
 'f0bb9e5a-eb11',
 '94c4ece8-bba6',
 '5ce60722-dbd1',
 '56126c14-b5b8',
 '83d174e1-5f9a',
 '355fe35a-5db7',
 'bdb3b33c-83a9',
 '06f377a6-ba7e',
 '77222291-e191',
 '1b22b0ab-aba3',
 'f86688f1-8742',
 '73ed1381-3a3f',
 '2618fdc4-e731',
 '7dbc4ebf

In [76]:
# verificando os valores das características dos IDs duplicados
df.loc[df['ID'].isin(dupe_ids[0:3]),:]

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
5033   89f8f447-fca8     320000    2          2         1   32     0      0   
5133   89f8f447-fca8          0    0          0         0    0     0      0   
16727  1fb3e3e6-a68d      80000    1          2         2   33     2      2   
16827  1fb3e3e6-a68d          0    0          0         0    0     0      0   
29685  ad23fe5c-7b09      50000    1          3         1   32     0      0   
29785  ad23fe5c-7b09          0    0          0         0    0     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
5033       0      0  ...     169371     172868     150827      8000      8000   
5133       0      0  ...          0          0          0         0         0   
16727      0      0  ...      27394      29922      31879         0      2000   
16827      0      0  ...          0          0          0         0         0   
29685      0      0  ...      12882       8131       3983      3000      2871   
29785      0      0  ...          0          0          0         0         0   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
5033       5500      6100      6000      5000                           0  
5133          0         0         0         0                           0  
16727      2000      3000      2600         0                           1  
16827         0         0         0         0                           0  
29685      1000       163      3983      3771                           1  
29785         0         0         0         0                           0  

[6 rows x 25 columns]

Vamos encontrar as linhas que somente tem zeros nas colunas (exceto coluna ID)

VAMOS **CRIAR UMA MATRIZ BOOLEANA** COM O **MESMO TAMANHO DO DATAFRAME**

------> USANDO A CONDIÇÃO É == 0

In [77]:
df_zero_mask = df == 0

In [78]:
df_zero_mask

ID  LIMIT_BAL    SEX  EDUCATION  MARRIAGE    AGE  PAY_1  PAY_2  \
0      False      False  False      False     False  False  False  False   
1      False      False  False      False     False  False  False  False   
2      False      False  False      False     False  False   True   True   
3      False      False  False      False     False  False   True   True   
4      False      False  False      False     False  False  False   True   
...      ...        ...    ...        ...       ...    ...    ...    ...   
29995  False      False  False      False     False  False   True   True   
29996  False      False  False      False     False  False  False  False   
29997  False      False  False      False     False  False  False  False   
29998  False      False  False      False     False  False  False  False   
29999  False      False  False      False     False  False   True   True   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0      False  False  ...       True       True       True      True     False   
1       True   True  ...      False      False      False      True     False   
2       True   True  ...      False      False      False     False     False   
3       True   True  ...      False      False      False     False     False   
4      False   True  ...      False      False      False     False     False   
...      ...    ...  ...        ...        ...        ...       ...       ...   
29995   True   True  ...      False      False      False     False     False   
29996  False  False  ...      False      False       True     False     False   
29997  False  False  ...      False      False      False      True      True   
29998   True   True  ...      False      False      False     False     False   
29999   True   True  ...      False      False      False     False     False   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0          True      True      True      True                       False  
1         False     False      True     False                       False  
2         False     False     False     False                        True  
3         False     False     False     False                        True  
4         False     False     False     False                        True  
...         ...       ...       ...       ...                         ...  
29995     False     False     False     False                        True  
29996     False     False      True      True                        True  
29997     False     False     False     False                       False  
29998     False     False     False     False                       False  
29999     False     False     False     False                       False  

[30000 rows x 25 columns]

AGORA PRECISAMOS **INDEXAR A MATRIZ BOOLEANA -> df_zero_mask** USANDO A INDEXAÇÃO DE INTEIROS (.iloc)

In [79]:
feature_zero_mask = df_zero_mask.iloc[:,1:].all(axis=1)

In [80]:
feature_zero_mask

0        False
1        False
2        False
3        False
4        False
         ...  
29995    False
29996    False
29997    False
29998    False
29999    False
Length: 30000, dtype: bool

In [81]:
sum(feature_zero_mask)

315

315 linhas tem 0 em todas as colunas

------ **APAGANDO** AS LINHAS COM ZEROS EM TODAS AS COLUNAS -> ELIMINANDO OS IDS DUPLICADOS --------

In [82]:
df_clean_1 = df.loc[~feature_zero_mask,:].copy()

ASSIM **SELECIONAMOS** ATRAVÉS DO (.loc) TODAS AS **LINHAS QUE NÃO POSSUEM ZEROS EM TODAS AS COLUNAS**

Criando uma cópia ao invés de ver a saída do DataFrame original

In [83]:
# verificamos a cópia

df_clean_1

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0      798fc410-45c1      20000    2          2         1   24     2      2   
1      8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2      85698822-43f5      90000    2          2         2   34     0      0   
3      0737c11b-be42      50000    2          2         1   37     0      0   
4      3b7f77cc-dbc0      50000    1          2         1   57    -1      0   
...              ...        ...  ...        ...       ...  ...   ...    ...   
29995  ecff42d0-bdc6     220000    1          3         1   39     0      0   
29996  99d1fa0e-222b     150000    1          3         2   43    -1     -1   
29997  95cdd3e7-4f24      30000    1          2         2   37     4      3   
29998  00d03f02-04cd      80000    1          3         1   41     1     -1   
29999  15d69f9f-5ad3      50000    1          2         1   46     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0         -1     -1  ...          0          0          0         0       689   
1          0      0  ...       3272       3455       3261         0      1000   
2          0      0  ...      14331      14948      15549      1518      1500   
3          0      0  ...      28314      28959      29547      2000      2019   
4         -1      0  ...      20940      19146      19131      2000     36681   
...      ...    ...  ...        ...        ...        ...       ...       ...   
29995      0      0  ...      88004      31237      15980      8500     20000   
29996     -1     -1  ...       8979       5190          0      1837      3526   
29997      2     -1  ...      20878      20582      19357         0         0   
29998      0      0  ...      52774      11855      48944     85900      3409   
29999      0      0  ...      36535      32428      15313      2078      1800   

       PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0             0         0         0         0                           1  
1          1000      1000         0      2000                           1  
2          1000      1000      1000      5000                           0  
3          1200      1100      1069      1000                           0  
4         10000      9000       689       679                           0  
...         ...       ...       ...       ...                         ...  
29995      5003      3047      5000      1000                           0  
29996      8998       129         0         0                           0  
29997     22000      4200      2000      3100                           1  
29998      1178      1926     52964      1804                           1  
29999      1430      1000      1000      1000                           1  

[29685 rows x 25 columns]

In [84]:
# agora verificamos se o número de linhas restantes é igual ao número de IDs exclusivos

# primeiro o número de linhas restantes

df_clean_1.shape

(29685, 25)

In [85]:
# agora o número de IDs exclusivos

df_clean_1['ID'].nunique()

29685

In [86]:
# IDs duplicados exclídos com sucesso!

In [88]:
dflimpo = pd.DataFrame(data = df_clean_1)

In [89]:
print(dflimpo)

                  ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0      798fc410-45c1      20000    2          2         1   24     2      2   
1      8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2      85698822-43f5      90000    2          2         2   34     0      0   
3      0737c11b-be42      50000    2          2         1   37     0      0   
4      3b7f77cc-dbc0      50000    1          2         1   57    -1      0   
...              ...        ...  ...        ...       ...  ...   ...    ...   
29995  ecff42d0-bdc6     220000    1          3         1   39     0      0   
29996  99d1fa0e-222b     150000    1          3         2   43    -1     -1   
29997  95cdd3e7-4f24      30000    1          2         2   37     4      3   
29998  00d03f02-04cd      80000    1          3         1   41     1     -1   
29999  15d69f9f-5ad3      50000    1          2         1   46     0      0   

       PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BIL

In [94]:
dflimpo.to_excel('dflimpo.csv')

ValueError: No engine for filetype: 'csv'